New Tuya camera support

seydx opened the pull request:
This PR adds comprehensive support for Tuya cameras in go2rtc, inspired by the [previous PR](https://github.com/AlexxIT/go2rtc/pull/1379) that introduced initial Tuya integration.

## Key Improvements

* Implemented `tuya://` protocol instead of `webrtc:` for better clarity and separation of concerns
* Added RTSP support for compatible cameras (enabled via `use_rtsp=1` parameter)
* Intelligent backchannel support that's automatically added based on camera capabilities/skills
* Code optimizations and bug fixes throughout the implementation
* Proper handling of multiple concurrent streams
* Clean resource management and proper cleanup on stream termination

This implementation has been tested with several Tuya camera models and provides a more robust experience while maintaining all functionality from the original PR.

## Usage Examples

```yaml
streams:
  # Basic Tuya camera stream
  tuya1: tuya://openapi.tuyaus.com?device_id=XXX&uid=XXX&client_id=XXX&secret=XXX
  
  # Using RTSP when available
  tuya2: tuya://openapi.tuyaus.com?device_id=XXX&uid=XXX&client_id=XXX&secret=XXX&use_rtsp=1

felipecrs commented:
I got this error when trying it:

```
panic: runtime error: index out of range [0] with length 0

goroutine 43 [running]:
github.com/AlexxIT/go2rtc/pkg/tuya.(*TuyaClient).InitDevice(0xc00024c400)
        github.com/AlexxIT/go2rtc/pkg/tuya/api.go:305 +0x885
github.com/AlexxIT/go2rtc/pkg/tuya.NewTuyaClient({0xc0001ea487, 0x12}, {0xc0001ea4a4, 0x16}, {0xc0001ea4bf, 0x14}, {0xc0001ea4de, 0x14}, {0xc0001ea4fa, 0x20}, ...)
        github.com/AlexxIT/go2rtc/pkg/tuya/api.go:169 +0x2bd
github.com/AlexxIT/go2rtc/pkg/tuya.Dial({0xc0001ea480, 0xb3})
        github.com/AlexxIT/go2rtc/pkg/tuya/client.go:51 +0x379
main.main.Init.func5({0xc0001ea480?, 0xc0001c42a0?})
        github.com/AlexxIT/go2rtc/internal/tuya/tuya.go:11 +0x18
github.com/AlexxIT/go2rtc/internal/streams.GetProducer({0xc0001ea480, 0xb3})
        github.com/AlexxIT/go2rtc/internal/streams/handlers.go:49 +0xcc
github.com/AlexxIT/go2rtc/internal/streams.(*Producer).Dial(0xc0000ec630)
        github.com/AlexxIT/go2rtc/internal/streams/producer.go:62 +0x8a
github.com/AlexxIT/go2rtc/internal/streams.(*Stream).AddConsumer(0xc0001d4300, {0xee4508, 0xc00023c120})
        github.com/AlexxIT/go2rtc/internal/streams/add_consumer.go:36 +0x676
github.com/AlexxIT/go2rtc/internal/mp4.handlerWSMSE(0xc0005b8000, 0xc0005ba000)
        github.com/AlexxIT/go2rtc/internal/mp4/ws.go:29 +0x145
github.com/AlexxIT/go2rtc/internal/api/ws.apiWS.func2()
        github.com/AlexxIT/go2rtc/internal/api/ws/ws.go:134 +0x39
created by github.com/AlexxIT/go2rtc/internal/api/ws.apiWS in goroutine 52
        github.com/AlexxIT/go2rtc/internal/api/ws/ws.go:133 +0x357
```

The config:

```yaml
streams:
  rua_tuya:
    - tuya://openapi.tuyaus.com?device_id=x&uid=x&client_id=x&secret=x
```

Maybe it's something related to my camera specifically, but it shouldn't panic anyway.

felipecrs commented:
PS: I got it working for another camera.

seydx commented:
> I got this error when trying it:
> 
> ```
> panic: runtime error: index out of range [0] with length 0
> 
> goroutine 43 [running]:
> github.com/AlexxIT/go2rtc/pkg/tuya.(*TuyaClient).InitDevice(0xc00024c400)
>         github.com/AlexxIT/go2rtc/pkg/tuya/api.go:305 +0x885
> github.com/AlexxIT/go2rtc/pkg/tuya.NewTuyaClient({0xc0001ea487, 0x12}, {0xc0001ea4a4, 0x16}, {0xc0001ea4bf, 0x14}, {0xc0001ea4de, 0x14}, {0xc0001ea4fa, 0x20}, ...)
>         github.com/AlexxIT/go2rtc/pkg/tuya/api.go:169 +0x2bd
> github.com/AlexxIT/go2rtc/pkg/tuya.Dial({0xc0001ea480, 0xb3})
>         github.com/AlexxIT/go2rtc/pkg/tuya/client.go:51 +0x379
> main.main.Init.func5({0xc0001ea480?, 0xc0001c42a0?})
>         github.com/AlexxIT/go2rtc/internal/tuya/tuya.go:11 +0x18
> github.com/AlexxIT/go2rtc/internal/streams.GetProducer({0xc0001ea480, 0xb3})
>         github.com/AlexxIT/go2rtc/internal/streams/handlers.go:49 +0xcc
> github.com/AlexxIT/go2rtc/internal/streams.(*Producer).Dial(0xc0000ec630)
>         github.com/AlexxIT/go2rtc/internal/streams/producer.go:62 +0x8a
> github.com/AlexxIT/go2rtc/internal/streams.(*Stream).AddConsumer(0xc0001d4300, {0xee4508, 0xc00023c120})
>         github.com/AlexxIT/go2rtc/internal/streams/add_consumer.go:36 +0x676
> github.com/AlexxIT/go2rtc/internal/mp4.handlerWSMSE(0xc0005b8000, 0xc0005ba000)
>         github.com/AlexxIT/go2rtc/internal/mp4/ws.go:29 +0x145
> github.com/AlexxIT/go2rtc/internal/api/ws.apiWS.func2()
>         github.com/AlexxIT/go2rtc/internal/api/ws/ws.go:134 +0x39
> created by github.com/AlexxIT/go2rtc/internal/api/ws.apiWS in goroutine 52
>         github.com/AlexxIT/go2rtc/internal/api/ws/ws.go:133 +0x357
> ```
> 
> The config:
> 
> ```yaml
> streams:
>   rua_tuya:
>     - tuya://openapi.tuyaus.com?device_id=x&uid=x&client_id=x&secret=x
> ```
> 
> Maybe it's something related to my camera specifically, but it shouldn't panic anyway.

yeah it shouldn't panic, will take a look

felipecrs commented:
Can you please also add an example on how to change the resolution to your PR description?

seydx commented:
> Can you please also add an example on how to change the resolution to your PR description?

done

seydx commented:
@felipecrs 

For the camera where go2rtc crashes with a panic, there seems to be no "skill" entry in the response from your camera

https://github.com/seydx/go2rtc/blob/ff00484328d2e2cf3f75dea4616529f52dbf356d/pkg/tuya/api.go#L305

It looks like I need to rewrite this somewhat so that it falls back to default video/audio.

felipecrs commented:
Thank you! I guess my camera doesn't support changing the resolution through the WebRTC API because regardless of `resolution=0` or `resolution=1` I always get the SD stream.

---

Some other feedback:

1. Supposing you're using the same API [I am using in my script](https://github.com/felipecrs/hass-expose-camera-stream-source/blob/master/custom_components/expose_camera_stream_source/scripts/get_tuya_stream_url.py) when `use_rtsp=1`, you should not need the user to pass `uid`.
2. Maybe you can simplify `use_rtsp=1` into `rtsp=1`
3. It would be great if you could also allow `use_hls=1` or `hls=1`. Some cameras only work through the HLS stream. It should be as simple as passing `"type": "HLS"` instead of `"type": "RTSP"` when requesting the stream ([reference from my script](https://github.com/felipecrs/hass-expose-camera-stream-source/blob/cf4627df2ce1b9eef6c7b5af1c653bf0eb977732/custom_components/expose_camera_stream_source/scripts/get_tuya_stream_url.py#L43)).

felipecrs commented:
> @felipecrs
> 
> For the camera where go2rtc crashes with a panic, there seems to be no "skill" entry in the response from your camera
> 
> [seydx/go2rtc@`ff00484`/pkg/tuya/api.go#L305](https://github.com/seydx/go2rtc/blob/ff00484328d2e2cf3f75dea4616529f52dbf356d/pkg/tuya/api.go#L305)
> 
> It looks like I need to rewrite this somewhat so that it falls back to default video/audio.

Do you want me to run some Tuya API against my camera? I can do it relatively easy through the Tuya IOT Debug function.

seydx commented:
> > For the camera where go2rtc crashes with a panic, there seems to be no "skill" entry in the response from your camera
> > [seydx/go2rtc@`ff00484`/pkg/tuya/api.go#L305](https://github.com/seydx/go2rtc/blob/ff00484328d2e2cf3f75dea4616529f52dbf356d/pkg/tuya/api.go#L305)
> > It looks like I need to rewrite this somewhat so that it falls back to default video/audio.
> 
> Do you want me to run some Tuya API against my camera? I can do it relatively easy through the Tuya IOT Debug function.

Yeah that would be nice, this is the endpoint:

https://github.com/seydx/go2rtc/blob/ff00484328d2e2cf3f75dea4616529f52dbf356d/pkg/tuya/api.go#L258

should look like:

```json
{
    "result":{
        "audio_attributes":{
            "call_mode":[
                1,
                2
            ],
            "hardware_capability":[
                1,
                2
            ]
        },
        "auth":"h85L4pljbuHFR0a/iTgViwA35xi3yTl3NyMsFQL5****",
        "id":"6cf2b6d2b09a2f8597****",
        "moto_id":"moto_cnpre002",
        "p2p_config":{
            "ices":[
                {
                    "urls":"stun:49.234.141.77:3478"
                },
                {
                    "urls":"stun:tx1stun.tuyacn.com:3478"
                },
                {
                    "urls":"nat:tx1nat.tuyacn.com:3478"
                },
                {
                    "urls":"nat:tx2nat.tuyacn.com:3478"
                },
                {
                    "credential":"kb/EA2whGCcNSM5FjXV2dxAM1MU=",
                    "ttl":36000,
                    "urls":"turn:49.234.141.77:3478",
                    "username":"1600883205:6cf2b6d2b09a2f8597****"
                },
                {
                    "credential":"kb/EA2whGCcNSM5FjXV2dxAM****",
                    "ttl":36000,
                    "urls":"turn:tx1turn.tuyacn.com:3478",
                    "username":"1600883205:6cf2b6d2b09a2f8597****"
                }
            ]
        },
        "skill":"{"webrtc":3,"audios":[{"channels":1,"dataBit":16,"codecType":101,"sampleRate":8000}],"videos":[{"streamType":2,"profileId":"","width":1920,"codecType":2,"sampleRate":90000,"height":1080},{"streamType":4,"width":640,"codecType":2,"height":360}]}",
        "supports_webrtc":true,
        "video_clarity":4
    },
    "success":true,
    "t":1600847205437
}

```

felipecrs commented:
This is the output:

```json
{
  "result": {
    "audio_attributes": {
      "call_mode": [
        1
      ],
      "hardware_capability": [
        1,
        2
      ]
    },
    "auth": "xx=",
    "id": "xx",
    "moto_id": "signaling14723",
    "p2p_config": {
      "auth": "xx=",
      "ices": [
        {
          "urls": "stun:44.240.73.31:3478"
        },
        {
          "urls": "stun:[2600:1f14:305f:5a01:3a96:c309:6fd1:4847]:3478"
        },
        {
          "credential": "xx",
          "ttl": 36000,
          "urls": "turn:15.204.9.11:3478",
          "username": "1747054638:eb05a8c8501ca6b9784gfw"
        }
      ],
      "moto_id": "signaling14723"
    },
    "protocol_version": "2.2",
    "skill": "{\"webrtc\":3}",
    "support_webrtc_record": false,
    "supports_webrtc": true,
    "vedio_clarity": 2,
    "vedio_claritys": [
      2,
      4
    ],
    "video_clarity": 2
  },
  "success": true,
  "t": 1747018638601,
  "tid": "d1178ccb2edc11f0b3586285baec33dd"
}
```

felipecrs commented:
Out of curiosity, this is the output for the camera which only works through HLS:

```json
{
  "result": {
    "audio_attributes": {
      "call_mode": [],
      "hardware_capability": [
        1
      ]
    },
    "auth": "BlLsgEVlVIpSH4KcSA6+hiLxZukf6S0WyUvPImUTPXA=",
    "id": "ebf031220b3de54904xgqs",
    "moto_id": "signaling14724",
    "p2p_config": {
      "auth": "BlLsgEVlVIpSH4KcSA6+hiLxZukf6S0WyUvPImUTPXA=",
      "ices": [
        {
          "urls": "stun:54.148.238.22:3478"
        },
        {
          "urls": "stun:[2600:1f14:305f:5a01:def3:5ab9:10ea:974e]:3478"
        },
        {
          "credential": "wh7Ux6mZ7jpFL5F0E4Cw/XGq540=",
          "ttl": 36000,
          "urls": "turn:15.204.8.186:3478",
          "username": "1747054814:ebf031220b3de54904xgqs"
        }
      ],
      "moto_id": "signaling14724"
    },
    "protocol_version": "2.2",
    "skill": "{\"webrtc\":2}",
    "support_webrtc_record": false,
    "supports_webrtc": true,
    "vedio_clarity": 4,
    "vedio_claritys": [
      2,
      4
    ],
    "video_clarity": 4
  },
  "success": true,
  "t": 1747018814643,
  "tid": "3a0a92932edd11f0bc8ac69aa343f8ac"
}
```

PS: It's a different camera, but it also panics in the same place.

felipecrs commented:
Out of curiosity, this is the output for the camera which only works through HLS:

```json
{
  "result": {
    "audio_attributes": {
      "call_mode": [],
      "hardware_capability": [
        1
      ]
    },
    "auth": "xx",
    "id": "xx",
    "moto_id": "signaling14724",
    "p2p_config": {
      "auth": "xx",
      "ices": [
        {
          "urls": "stun:54.148.238.22:3478"
        },
        {
          "urls": "stun:[2600:1f14:305f:5a01:def3:5ab9:10ea:974e]:3478"
        },
        {
          "credential": "xx",
          "ttl": 36000,
          "urls": "turn:15.204.8.186:3478",
          "username": "1747054814:ebf031220b3de54904xgqs"
        }
      ],
      "moto_id": "signaling14724"
    },
    "protocol_version": "2.2",
    "skill": "{\"webrtc\":2}",
    "support_webrtc_record": false,
    "supports_webrtc": true,
    "vedio_clarity": 4,
    "vedio_claritys": [
      2,
      4
    ],
    "video_clarity": 4
  },
  "success": true,
  "t": 1747018814643,
  "tid": "3a0a92932edd11f0bc8ac69aa343f8ac"
}
```

PS: It's a different camera, but it also panics in the same place.

seydx commented:
@felipecrs 

panic should be fixed, can you try pls, change also the rtsp query to `rtsp` and added a new hls query `hls`

felipecrs commented on pkg/tuya/api.go:
```diff
@@ -162,9 +163,13 @@ func NewTuyaClient(openAPIURL string, deviceID string, uid string, clientID stri
+		if err := client.GetStreamUrl("rtsp"); err != nil {
 			return nil, fmt.Errorf("failed to get RTSP URL: %w", err)
 		}
+	} else if useHLS {
+		if err := client.GetStreamUrl("hls"); err != nil {
+			return nil, fmt.Errorf("failed to get HLS URL: %w", err)
+		}
```
I think this fallback mechanism is not useful. In my tests, the RTSP stream URL is always returned, but the stream itself doesn't work when played.

felipecrs replied:
Oops, never mind. I misread the code.

felipecrs commented:
> panic should be fixed, can you try pls, changed also the rtsp query to `rtsp` and added a new hls query `hls`

Not only the panic is gone, I can now stream that camera over WebRTC. Thanks!!

seydx commented:
> > panic should be fixed, can you try pls, changed also the rtsp query to `rtsp` and added a new hls query `hls`
> 
> Not only the panic is gone, I can now stream that camera over WebRTC. Thanks!!

Tested HLS? HLS does not work for my camera, have only black screen. Using ffmpeg as internal producer with disabled audio, because with audio the stream was not playable

felipecrs commented:
Wow, the stream for my camera which previously only worked through HLS is now possible with WebRTC too!

felipecrs commented:
The HLS mode is kinda weird though:

![image](https://github.com/user-attachments/assets/fd197df9-8304-4bc3-8731-234bc45861b9)

felipecrs commented:
Honestly I will probably never need to use HLS anymore, since WebRTC is working. HLS is such a mess.

seydx commented:
> The HLS mode is kinda weird though:
> 
> 
> 
> ![image](https://github.com/user-attachments/assets/fd197df9-8304-4bc3-8731-234bc45861b9)
> 
> 

Ffmpeg version? It seems it doesnt support https protocol

felipecrs commented:
> Ffmpeg version? It seems it doesnt support https protocol

Yeah, that's probably it, I'm using it through [pkgx](https://github.com/pkgxdev/pkgx#readme). Probably they built it without HTTPS support.

felipecrs commented:
> can you show me the go2rtc stream config related to the camera pls

```yaml
streams:
  bebe:
    - tuya://openapi.tuyaus.com?device_id=xx&uid=xx&client_id=xx&secret=xx&hls=1
```

felipecrs commented:
> because with audio the stream was not playable

Yeah, looks like go2rtc doesn't play well with ADTS audio.

felipecrs commented:
If I were you, I'd remove `ffmpeg` from the HLS stream and just let it be. The problem is that go2rtc cannot play it, and maybe someday it will.

I had the same issue when using my script. Was never able to get the `.m3u8` HLS stream working in go2rtc, but in VLC it plays (it just takes some seconds in the black loading screen).

BTW I tried your PR in my Frigate instance, with a proper ffmpeg, and I was able to get the black screen, just like I used to do using my script in HLS mode in go2rtc.

So, again, I think your PR is doing what it's supposed to for HLS mode. The rest is up to go2rtc itself.

felipecrs commented:
> > because with audio the stream was not playable
> 
> Yeah, looks like go2rtc doesn't play well with ADTS audio.

Actually I cannot reproduce that. Maybe your camera outputs a different audio codec when in HLS mode. For me the stream plays the same way with or without `ffmpeg`.

felipecrs commented:
@seydx thinking better about the parameters, I think the following is what makes the most sense:

Instead of `rtsp=1` or `hls=1`, you can have `type=rtsp`, `type=hls`, or `type=webrtc`, where if none is set, `type=webrtc` is assumed.

seydx commented:
> @seydx thinking more about the parameters, I think the following is what makes the most sense:
> 
> Instead of `rtsp=1` or `hls=1`, you can have `type=rtsp`, `type=hls`, or `type=webrtc`, where if none is set, `type=webrtc` is assumed.

done

seydx commented:
> If I were you, I'd remove `ffmpeg` from the HLS stream and just let it be. The problem is that go2rtc cannot play it, and maybe someday it will.
> 
> I had the same issue when using my script. Was never able to get the `.m3u8` HLS stream working in go2rtc, but in VLC it plays (it just takes some seconds in the black loading screen).
> 
> BTW I tried your PR in my Frigate instance, with a proper ffmpeg, and I was able to get the black screen, just like I used to do using my script in HLS mode in go2rtc.
> 
> So, again, I think your PR is doing what it's supposed to for HLS mode. The rest is up to go2rtc itself.

done

felipecrs commented:
Not sure if it's worth it, probably not, but you could get the user id using `client_id` and `client_secret` through this API, and looking for the first user returned:

https://developer.tuya.com/en/docs/archived-documents/user-management?id=K95ztzvgwnshy#title-9-Get%20the%20user%20list

felipecrs commented:
Another recommendation I would give is to name the `secret` parameter as `client_secret`.

`client_secret` makes a better direct relationship to how it is named in Tuya IOT, and also this is how it is called in other interfaces like Home Assistant and Local Tuya:

- https://github.com/home-assistant/core/blob/b192ca4bad73b16f9d313277d4d92d5a5201694e/homeassistant/const.py#L141-L142
- https://github.com/xZetsubou/hass-localtuya/blob/c4923fc68352ba096f9c23ca276de2e0c7d4ee2b/custom_components/localtuya/config_flow.py#L27-L28

![image](https://github.com/user-attachments/assets/4d88f1d0-791c-4a6b-b835-b4508345a867)

I guess keeping the same terminology across different interfaces helps increase predictability and reduce confusion.

seydx commented:
> Not sure if it's worth it, probably not, but you could get the user id using `client_id` and `client_secret` through this API, and looking for the first user returned:
> 
> https://developer.tuya.com/en/docs/archived-documents/user-management?id=K95ztzvgwnshy#title-9-Get%20the%20user%20list

Not tested yet but are you sure that it works?

![IMG_3983](https://github.com/user-attachments/assets/ad50a675-7c82-49de-875f-9741e9e18160)

felipecrs commented on pkg/tuya/mqtt.go:
```diff
@@ -0,0 +1,273 @@
+
+	payload, err := json.Marshal(msg)
+	if err != nil {
+		c.mqtt.onError(fmt.Errorf("marshal mqtt message fail: %s", err.Error()))
```
Maybe you should prefix all errors with something like `tuya: `. I don't know how it's done for other modules though.

felipecrs replied:
Also, I could not find out where, but you are printing the RTSP or HLS URL in the logs.

It certainly helps troubleshooting, but I think it should be printed only in DEBUG level.

Speaking of it, I think you need to refactor all logs to use the logger module.

seydx replied:
Fixed

felipecrs commented:
> Not tested yet but are you sure that it works?

Oh, I thought it was only redacted for the example. Yeah, that's not good. Also, looks like it needs `{schema}` which seems to be the name of the project in Tuya IOT, which would require user input anyway. Ok, never mind about it. Sorry.

felipecrs commented:
@seydx but you still should not need user id when `type=rtsp` or `type=hls`, right? Maybe you can alleviate the query validation when `type=rtsp` or `type=hls`.

felipecrs commented:
Also, do you think there's any way to validate the `resolution` parameter based on the camera capabilities?

It would be nice to fail early if user tries to use it when not compatible, instead of having to guess whether it worked or not.

Looks like your camera supports it:

```json
"skill":"{"webrtc":3,"audios":[{"channels":1,"dataBit":16,"codecType":101,"sampleRate":8000}],"videos":[{"streamType":2,"profileId":"","width":1920,"codecType":2,"sampleRate":90000,"height":1080},{"streamType":4,"width":640,"codecType":2,"height":360}]}",
```

But all the cameras I have don't:

```json
"skill": "{\"webrtc\":3}",
```

```
"skill": "{\"webrtc\":2}",
```

felipecrs commented:
> Tested HLS? HLS does not work for my camera, have only black screen. Using ffmpeg as internal producer with disabled audio, because with audio the stream was not playable

If you want to keep troubleshooting the HLS stream, can you check whether it works in VLC for you (like it does for me)?

I'm not sure if two clients can connect to the same stream at the same time, so maybe a safer bet is to use my script to output the stream:

https://github.com/felipecrs/hass-expose-camera-stream-source#bonus-importing-tuya-cameras-to-go2rtc-without-home-assistant

felipecrs commented:
> Its possible, but i am not sure if the value for `skill` is required for `resolution` to work. My camera has two streams as you see, H265 (streamType = 2) and H264 (streamType = 4). But i can not change it via `resolution` query. Tuya can only use the H264 stream

1. Right, if unsure, better not prevent users from using it then. We may be preventing it from working in unknown situations.
2. Does it mean you were not able to test `resolution` at all? (or did you mean `resolution` doesn't work for that specific camera you have, and you were able to test it with some other camera?)
   
   Reason I ask is: if you were unable to test it at all, there's a chance it's not working because of some missing or incorrect logic.

seydx commented:
> > Its possible, but i am not sure if the value for `skill` is required for `resolution` to work. My camera has two streams as you see, H265 (streamType = 2) and H264 (streamType = 4). But i can not change it via `resolution` query. Tuya can only use the H264 stream
> 
> 1. Right, if unsure, better not prevent users from using it then. We may be preventing it from working in unknown situations.
> 2. Does it mean you were not able to test `resolution` at all? (or did you mean `resolution` doesn't work for that specific camera you have, and you were able to test it with some other camera?)
>    Reason I ask is: if you were unable to test it at all, there's a chance it's not working because of some missing or incorrect logic.
>    PS: All my cameras have two streams too, SD and HD, and I can swap them through the Tuya app. They are all H264.

I have only cameras with one h265 stream and one h264 sub stream, so for me
Its not working. Tuya sends always the sub stream because of h264.

Is it possible to get access to one of your cameras to test ?

felipecrs commented:
Yes, but how? Can you ping me at https://felipecrs.com/telegram or https://felipecrs.com/discord?

seydx commented:
> > Its possible, but i am not sure if the value for `skill` is required for `resolution` to work. My camera has two streams as you see, H265 (streamType = 2) and H264 (streamType = 4). But i can not change it via `resolution` query. Tuya can only use the H264 stream
> 
> 1. Right, if unsure, better not prevent users from using it then. We may be preventing it from working in unknown situations.
> 2. Does it mean you were not able to test `resolution` at all? (or did you mean `resolution` doesn't work for that specific camera you have, and you were able to test it with some other camera?)
>    Reason I ask is: if you were unable to test it at all, there's a chance it's not working because of some missing or incorrect logic.
>    PS: All my cameras have two streams too, SD and HD, and I can swap them through the Tuya app. They are all H264.

removed the `resolution` logic and changed the query parameters a bit. Stream type (main/sub) is now available via `type` query

felipecrs commented:
I liked these changes a lot!

Please change the PR description from `secret` to `client_secret` when you can, I think you forgot it.

seydx commented:
@AlexxIT 

PR is good to go. Give it a look when you can, no rush.

felipecrs commented:
@pergolafabio you should be able to view H265 through WebRTC in Chrome 136+ btw.

felipecrs commented:
> @AlexxIT
> 
> PR is good to go. Give it a look when you can, no rush.

I can confirm everything is working great with this PR. It's going to be a game changer.

pergolafabio commented:
> > hey; was also testing this, but using rtsp source, since my camera is h265 , and that doesnt work it seems with webrtc anyway,, used this, changed to "eu"
> > `tuya_rtsp: tuya://openapi.tuyaeu.com?device_id=bfa8726443807xxx&client_id=xf3tjqp5ggaxxxx&client_secret=cfd41d4235eb467eafxxxxxxxxxx&mode=rtsp ` But i get :
> > ![image](https://private-user-images.githubusercontent.com/43251136/443213684-a946ac71-e28c-4205-8fcd-ba60765255f5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDcxNDE0NjcsIm5iZiI6MTc0NzE0MTE2NywicGF0aCI6Ii80MzI1MTEzNi80NDMyMTM2ODQtYTk0NmFjNzEtZTI4Yy00MjA1LThmY2QtYmE2MDc2NTI1NWY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTEzVDEyNTkyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY3ZmU4MjAyNTVkMzU2M2I0YTgwMGI0NDljMzIxYzhkYmQ0MDdkYzMwOGY2M2E1M2VhOTEyMjlhYTJhMDQ4MmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9r4XjmS8j6jqJSe7CqV7RKMyPZE8uhN6m7cJl2aKnXk)
> > are those "\u0026" normal in that path ?
> > ![image](https://private-user-images.githubusercontent.com/43251136/443214106-6674c73d-cba8-41af-b6fd-dc8ad6798568.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDcxNDE0NjcsIm5iZiI6MTc0NzE0MTE2NywicGF0aCI6Ii80MzI1MTEzNi80NDMyMTQxMDYtNjY3NGM3M2QtY2JhOC00MWFmLWI2ZmQtZGM4YWQ2Nzk4NTY4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTEzVDEyNTkyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxYTJkOTMzOWY0Zjc4OWFiMTA2YmM5YmY5ZmJmYmUzMDRkZjU2ZmJkNzBlYzY0MzgzOTU3ZjkyOWFhN2E0NTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.cdzovn7RJa-kBHQDd92gmcyh5lNLneNMBCYGlFuu3pc)
> 
> fixed, try again pls



Hey, its indeed now working with latest fix!! rtsp works
, i was now testing webrtc, there i still see:

![image](https://github.com/user-attachments/assets/97faa4b3-299c-49fe-8107-03f8f1016239)

pergolafabio commented:
> @pergolafabio you should be able to view H265 through WebRTC in Chrome 136+ btw.

i still see black screen with webrtc and chrome 136 :-), is https needed ? just testing the binary on http://ip:1984

seydx commented:
> > > hey; was also testing this, but using rtsp source, since my camera is h265 , and that doesnt work it seems with webrtc anyway,, used this, changed to "eu"
> > > `tuya_rtsp: tuya://openapi.tuyaeu.com?device_id=bfa8726443807xxx&client_id=xf3tjqp5ggaxxxx&client_secret=cfd41d4235eb467eafxxxxxxxxxx&mode=rtsp ` But i get :
> > > ![image](https://private-user-images.githubusercontent.com/43251136/443213684-a946ac71-e28c-4205-8fcd-ba60765255f5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDcxNDE0NjcsIm5iZiI6MTc0NzE0MTE2NywicGF0aCI6Ii80MzI1MTEzNi80NDMyMTM2ODQtYTk0NmFjNzEtZTI4Yy00MjA1LThmY2QtYmE2MDc2NTI1NWY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTEzVDEyNTkyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY3ZmU4MjAyNTVkMzU2M2I0YTgwMGI0NDljMzIxYzhkYmQ0MDdkYzMwOGY2M2E1M2VhOTEyMjlhYTJhMDQ4MmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9r4XjmS8j6jqJSe7CqV7RKMyPZE8uhN6m7cJl2aKnXk)
> > > are those "\u0026" normal in that path ?
> > > ![image](https://private-user-images.githubusercontent.com/43251136/443214106-6674c73d-cba8-41af-b6fd-dc8ad6798568.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDcxNDE0NjcsIm5iZiI6MTc0NzE0MTE2NywicGF0aCI6Ii80MzI1MTEzNi80NDMyMTQxMDYtNjY3NGM3M2QtY2JhOC00MWFmLWI2ZmQtZGM4YWQ2Nzk4NTY4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTEzVDEyNTkyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxYTJkOTMzOWY0Zjc4OWFiMTA2YmM5YmY5ZmJmYmUzMDRkZjU2ZmJkNzBlYzY0MzgzOTU3ZjkyOWFhN2E0NTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.cdzovn7RJa-kBHQDd92gmcyh5lNLneNMBCYGlFuu3pc)
> > 
> > 
> > fixed, try again pls
> 
> Hey, its indeed now working with latest fix!! rtsp works , i was now testing webrtc, there i still see:
> 
> ![image](https://private-user-images.githubusercontent.com/43251136/443225827-97faa4b3-299c-49fe-8107-03f8f1016239.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDcxNDI3NzksIm5iZiI6MTc0NzE0MjQ3OSwicGF0aCI6Ii80MzI1MTEzNi80NDMyMjU4MjctOTdmYWE0YjMtMjk5Yy00OWZlLTgxMDctMDNmOGYxMDE2MjM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTEzVDEzMjExOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNmNjg1MmUwMDI4NjM2MzdmNTQ4NTBiYWMyYzg1ZjQ5MTJjNWY2YjI5ZDQ1YmM0YmJjN2M2MGY5MjNlYjQxMWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.D99RyFZs7Z3a-leawT0diSrutNSxXY-6TUjTRjE4Ye4)

Is it possible to get access to the camera?

pergolafabio commented:
can i share the camera to your tuya account? you can ping me on discord: "fabiopergola"

felipecrs commented:
@pergolafabio does your camera work through the IPC Terminal?

- https://ipc-eu.ismartlife.me/
- https://ipc-us.ismartlife.me/

pergolafabio commented:
> @pergolafabio does your camera work through the IPC Terminal?
> 
> * https://ipc-eu.ismartlife.me/
> * https://ipc-us.ismartlife.me/

yes!

pergolafabio commented:
> @pergolafabio does your camera work through the IPC Terminal?
> 
> * https://ipc-eu.ismartlife.me/
> * https://ipc-us.ismartlife.me/

correction, that link doesnt work => decode error
![image](https://github.com/user-attachments/assets/3adb1dd0-ea05-4160-9beb-da02c5ac8a09)

but this works

https://protect-eu.ismartlife.me/playback

felipecrs commented on README.md:
```diff
@@ -565,6 +566,41 @@ streams:
+
+[Tuya](https://www.tuya.com/) proprietary camera protocol with **two way audio** support.
+
+- Obtain `client_id`, `client_secret`, `uid` and `device_id` from [Tuya IoT Platform](https://iot.tuya.com/)
```
```suggestion
- Obtain `client_id`, `client_secret`, `device_id`, and `uid` (if using `mode=webrtc`) from [Tuya IoT Platform](https://iot.tuya.com/). [Here's a guide](https://xzetsubou.github.io/hass-localtuya/cloud_api/).
```

Just a suggestion, integrate it however you want.

felipecrs commented on README.md:
```diff
@@ -575,7 +575,7 @@ Tested: KD110, KC200, KC401, KC420WS, EC71.
   - `rtsp` - RTSP stream _(if available)_
   - `hls` - HLS stream _(if available)_
-- Use `type` parameter to select the stream type: _(if available)_
+- Use `role` parameter to select the stream:
```
Hey, you don't need to change this if you disagree.

But I'm a bit uncomfortable on the name of this parameter.

I think naming it `stream` (like `&stream=main`, `&stream=sub`) would have made it more descriptive. Or even `resolution`.

felipecrs replied:
Actually I see you already refer to stream in `mode` parameter, like _WebRTC RTSP stream_, _HLS stream_.

Maybe `resolution` is still better?

seydx replied:
resolution = high/low ? hd/sd ?

felipecrs replied:
Because hd/sd makes a more direct reference to Tuya interfaces (both the apps and the IPC terminal), I think I like it better.

felipecrs replied:
Off-topic, but this could be a good example of a source providing 2 streams and allowing users to select the stream based on rtsp query parameters, rather than having to have 2 streams predefined.

- https://github.com/AlexxIT/go2rtc/issues/174

felipecrs commented on README.md:
```diff
@@ -565,6 +566,62 @@ streams:
+  - `flv` - FLV stream _(only available for `Open API`)_
+  - `rtmp` - RTMP stream _(only available for `Open API`)_
+
+- Use `resolution` parameter to select the stream (only available for `Cloud API` and not all cameras support `hd` stream):
```
I can visualize people opening issues that their cameras support HD/SD but this is not working, I think we can make it clearer:

```suggestion
- Use `resolution` parameter to select the stream (only available for `Cloud API` and not all cameras support `hd` stream through WebRTC even if supported in the camera itself):
```

felipecrs commented:
@AlexxIT this is the first open-source project in the world to support Tuya H.265 streams over WebRTC, and also the first open-source project in the world to support 2-way audio for Tuya cameras besides the infamous [tuya/webrtc-demo-go](https://github.com/tuya/webrtc-demo-go).

seydx commented:
Please don't merge yet. I'm going to rework this PR. Through reverse engineering, I managed to massively simplify the authentication - now it only needs email and password! This would actually make go2rtc the first open source project that can add and stream Tuya cameras super easily.

felipecrs commented on README.md:
```diff
@@ -568,58 +568,38 @@ Tested: KD110, KC200, KC401, KC420WS, EC71.
 #### Source: Tuya
 
-[Tuya](https://www.tuya.com/) proprietary camera protocol with **two way audio** support. Go2rtc supports `Cloud API` and `Open API`.
+[Tuya](https://www.tuya.com/) proprietary camera protocol with **two way audio** support. Go2rtc supports `Cloud API` and `Tuya API`.
```
Would it be better named _Tuya Cloud API_ and _Tuya Smart API_?

felipecrs commented:
That's not how it works, @halloworld007. It will be merged when AlexxIT feels like merging it. There's nothing we can do about it.

halloworld007 commented:
yes howto can insert this tuya pull 

sorry i am new on github

pergolafabio commented:
Guys, just wait if you are not familiar to this

lfelipeserrano commented:
Not sure if this is the place to ask about this, but giving a try to this PR on my local to play around with tuya cameras, I got it working using the Tuya Smart API, but having problems with the Tuya Cloud API, I'm getting this error: "No permissions. This API is not subscribed".
However, I have this running on my HA using the script from @felipecrs so I think it should be working, or am I missing something here? 
Thanks!

seydx commented:
@lfelipeserrano 

yeah thats correct

felipecrs commented:
> Thank you, it's asking me to activate a "Free Trial", I did and it will expire in a year, not sure if this have some cost?

They never charged me, and I have it for like 3 years. I never entered a credit card either, so they couldn't even charge me.

felipecrs commented:
Another suggestion would be to rank Tuya Smart API first in the README, and add a `(recommended)` next to it, so that users don't go down the rabbit hole of setting up the Cloud project without need.

felipecrs commented on README.md:
````diff
@@ -565,6 +566,43 @@ streams:
+  - `hd` - HD stream (default)
+  - `sd` - SD stream
+
+```yaml
+streams:
+  # Tuya Cloud API: WebRTC main stream
+  tuya_webrtc:
+   - tuya://openapi.tuyaus.com?device_id=XXX&uid=XXX&client_id=XXX&client_secret=XXX
+  
+  # Tuya Cloud API: WebRTC sub stream
+  tuya_webrtc_sd:
+   - tuya://openapi.tuyaus.com?device_id=XXX&uid=XXX&client_id=XXX&client_secret=XXX&resolution=sd
+
+  # Tuya Smart API: WebRTC main stream
+  tuya_main:
+    - tuya://protect-us.ismartlife.me?device_id=XXX&email=XXX&password=XXX
+
+  # Tuya Smart API: WebRTC sub stream
+  tuya_sub:
+    - tuya://protect-us.ismartlife.me?device_id=XXX&email=XXX&password=XXX&resolution=sd
+```
````
```suggestion
```yaml
streams:
  # Tuya Smart API: WebRTC main stream (use Add > Tuya to discover the URL)
  tuya_main:
    - tuya://protect-us.ismartlife.me?device_id=XXX&email=XXX&password=XXX

  # Tuya Smart API: WebRTC sub stream (use Add > Tuya to discover the URL)
  tuya_sub:
    - tuya://protect-us.ismartlife.me?device_id=XXX&email=XXX&password=XXX&resolution=sd

  # Tuya Cloud API: WebRTC main stream
  tuya_webrtc:
   - tuya://openapi.tuyaus.com?device_id=XXX&uid=XXX&client_id=XXX&client_secret=XXX
  
  # Tuya Cloud API: WebRTC sub stream
  tuya_webrtc_sd:
   - tuya://openapi.tuyaus.com?device_id=XXX&uid=XXX&client_id=XXX&client_secret=XXX&resolution=sd
```
```

kush-agra commented:
I was building something similar in python but this is so much better - thank you so much!!

for some reason in my code my rtsp stream returned from the api kept timing out every 7 minutes but the same doesn't happen in this code! cool!!

kush-agra commented:
Fail to refresh permissions: transaction closed - this is the error when that random 1 minute drop occurs 

and after 3-4 hours it decided to kill itself entirely - [logs](https://pastebin.com/P7c8FHmY)

seydx commented:
> Fail to refresh permissions: transaction closed - this is the error when that random 1 minute drop occurs
> 
> and after 3-4 hours it decided to kill itself entirely

should be fixed, can you try again pls with the new version

CLAM01 commented:
> > Thanks @seydx that got it working, is there any preference between the two methods? Like, any advantage from using Tuya Cloud API vs Tuya Smart API?
> 
> i prefere email/password, it does not require any cloud project, but its limited to tuya accounts only (smartlife accounts does not work)

Hi everyone, I came across this project and think it's great. I also have a Tuya Outdoor Cam and use Frigate.

I have loaded version 1.9.9 into Frigate as described and am now trying to get the stream via user and password.

But my camera is on a server in eu. For this reason I have used the following string:

go2rtc:
 streams:
      - tuya://protect-**eu**.ismartlife.me?device_id=xxx&email=xxx&password=xxx&resolution=sd

But I get a message from the Tuya server that the user names are wrong. But that can't be, because I can use them to get into the Tuya app.

What am I doing wrong, what do I have to provide for analysis?

seydx commented:
> 您的密码输入错误次数过多,请15分钟后再尝试

hm maybe parsing issue, can you try with a simple password

kush-agra commented:
I agree with the complexity not being worth it

CLAM01 commented:
> > > > Thanks @seydx that got it working, is there any preference between the two methods? Like, any advantage from using Tuya Cloud API vs Tuya Smart API?
> > > 
> > > 
> > > i prefere email/password, it does not require any cloud project, but its limited to tuya accounts only (smartlife accounts does not work)
> > 
> > 
> > Hi everyone, I came across this project and think it's great. I also have a Tuya Outdoor Cam and use Frigate.
> > I have loaded version 1.9.9 into Frigate as described ([#315 (comment)](https://github.com/AlexxIT/go2rtc/issues/315#issuecomment-2905955963)) and am now trying to get the stream via user and password. Here is the Log in Frigate: 2025-06-10 09:01:21.532160109 11:01:21.532 INF go2rtc platform=linux/amd64 revision=mod.d31afbe version=1.9.9-cui.20
> > But my camera is on a server in eu. For this reason I have used the following string:
> > go2rtc: streams: - tuya://protect-**eu**.ismartlife.me?device_id=xxx&email=xxx&password=xxx&resolution=sd
> > But I get a message from the Tuya server that the user names are wrong. But that can't be, because I can use them to get into the Tuya app.
> > Frigate Log: 2025-06-10 09:02:18.578945658 11:02:18.578 WRN [rtsp] error="streams: tuya: failed to initialize token: 您的密码输入错误次数过多,请15分钟后再尝试。" stream=tuya
> > What am I doing wrong, what do I have to provide for analysis?
> 
> maybe url parsing issue, can you try with a simpler password pls

Thanks a lot.
My Password conatins # and . (dot). After i delete this, the stream works.

seydx commented:
> > is there any preference between the two methods? Like, any advantage from using Tuya Cloud API vs Tuya Smart API?
> 
> I used the cloud api, for some reason the smart api process didn't work for me, tried both with us west account and India account both of which are the closest servers Error I received said this method is only supported with mainland China server
> 
> I already had cloud keys so i resorted to them

It doesn't depend on your location, it depends on the region you selected when you created the tuya account. Try the other regions, maybe `protect.ismartlife.me` as the endpoint for `China`

```go
var AvailableRegions = []Region{
	{"eu-central", "protect-eu.ismartlife.me", "Central Europe", "EU"},
	{"eu-east", "protect-we.ismartlife.me", "East Europe", "EU"},
	{"us-west", "protect-us.ismartlife.me", "West America", "AZ"},
	{"us-east", "protect-ue.ismartlife.me", "East America", "AZ"},
	{"china", "protect.ismartlife.me", "China", "AY"},
	{"india", "protect-in.ismartlife.me", "India", "IN"},
}
```

seydx commented:
> Thanks a lot. My Password conatins # and . (dot). After i delete this, the stream works.

fixed

felipecrs commented:
Yes, this is the format:

```
tuya://protect-us.ismartlife.me?device_id=eb05a8c8501ca6b9784gfw&email=youremail%40mail.com&password=yourpass
```

The device id is the same as the one from Tuya IOT, at least for me.

seydx commented:
> Oh yes that is correct, I meant I tried with accounts created on both of those server regions but none worked
> 
> I selected the corresponding region on go2rtc with which I created my account and in the server i created the account but no dice

oh ok, if you can give me your tuya account credentials, i can look at it

bantunes22 reviewed (approved):
LGTM.

I've also tried this with 3 Tuya Cameras via Tuya Smart App. And everything is working properly.

kush-agra commented:
> > Oh yes that is correct, I meant I tried with accounts created on both of those server regions but none worked
> > I selected the corresponding region on go2rtc with which I created my account and in the server i created the account but no dice
> 
> oh ok, if you can give me your tuya account credentials, i can look at it


sure!! how would you like me to give it to you? or we could create dummy creds in the same server because it won't even let you login in the first place

dwd1 commented:
seydx 1.9.9-cui.33 is not working for me (Tuya Cloud API, Tuya Smart API), but felipecrs addon worked before.
I only hear a click sound from the speaker every few seconds, so it's trying but there's no picture.

![image](https://github.com/user-attachments/assets/4edfcfe9-1d2f-404c-994e-f87cc1e6b170)

dwd1 commented:
My config is very simple:
 
```
streams:
  doorbell_rtsp:
    - tuya://openapi.tuyaeu.com?device_id=MY_DEVICE_ID&uid=MY_UID&client_id=MY_CLIENT_ID&client_secret=MY_CLIENT_SECRET
#    - tuya://protect-eu.ismartlife.me?device_id=MY_DEVICE_ID&email=MY_EMAIL&password=MY_PASSWORD

rtsp:
  listen: ":8554"
  username: ""
  password: ""
  default_query: "video&audio"
```

Its still working fine with 1.9.9-tuya.0 addon version installed from "AlexxIT addons repository (@felipecrs' fork)" [https://github.com/felipecrs/alexxit-hassio-addons].
Its NOT working with seydx 1.9.9-cui.34.

I dont know what other info can I say...

felipecrs commented:
@dwd1 don't you have anything related in logs?

grillead commented:
Hi, go2rtc with this PR work for me with my tuya doorbell but no sound , any config is needed ?
``` yaml
go2rtc:
  streams:
    Sonnette: #Tuya App
      - tuya://protect-eu.ismartlife.me?device_id=deviceID&email=email&password=password&resolution=sd
      #- ffmpeg:Sonnette#audio=copy#audio=opus
    Sonnette_HD: #Tuya App Hd
      - tuya://protect-eu.ismartlife.me?device_id=deviceID&email=email&password=password&resolution=hd
 ```
Thank's

seydx commented:
> Hi @seydx , I've been trying to use your PR to access the video stream of my birdfeeder camera. I can successfully connect to it via https://ipc-eu.ismartlife.me . But i'm having trouble pairing it with my tuya dev platform account. The problem is that the camera does not work with Tuya/SmartLife apps (but it used to work a few years ago), I need to use the Birdlover app instead. I think this is why I am unable to do the "linking app account". When I accept the QR code in the Birdlover app, the platform says that the QR code expired. Tried multiple times...
> 
> Do you have any advice on that?

Tuya account in required. It will even fail with a smartlife account

AlexxIT commented:
Well. This code don't work for my SmartLife account and for my Tuya account. I still don't know why.

seydx commented:
> > Well. This code don't work for my SmartLife account and for my Tuya account. I still don't know why.
> 
> I tested this with Tuya. Using the Tuya cloud has been working for about two weeks. Direct login with login/password does not work when the EU region is selected.

No issues for me with EU data center

p0is0n commented:
> > Hi, I tried using this PR, but I got the error:
> > `[rtsp] error="streams: tuya: failed to start MQTT: network Error : dial tcp 45.195.222.93:443: i/o timeout" stream=cam_1`
> > Connect via email/password region EU:
> > ```
> > tuya://protect-eu.ismartlife.me?device_id=XXX&email=XXX&password=XXX
> > ```
> 
> which country?

Kazakhstan. Ping to this IP is ok:

```
PING 45.195.222.93 (45.195.222.93): 56 data bytes
64 bytes from 45.195.222.93: icmp_seq=0 ttl=34 time=53.768 ms
64 bytes from 45.195.222.93: icmp_seq=1 ttl=34 time=56.417 ms
```

felipecrs commented:
Ping works but the 443 port doesn't seem open/accessible:

```console
~ ❯ ping 45.195.222.93
Pinging 45.195.222.93 with 32 bytes of data:
Reply from 45.195.222.93: bytes=32 time=358ms TTL=237

~ ❯ telnet 45.195.222.93 443
Connecting To 45.195.222.93...Could not open connection to the host, on port 443: Connect failed
```

From Brazil.

p0is0n commented:
> Ping works but the 443 port doesn't seem open/accessible:
> 
> ```
> ~ ❯ ping 45.195.222.93
> Pinging 45.195.222.93 with 32 bytes of data:
> Reply from 45.195.222.93: bytes=32 time=358ms TTL=237
> 
> ~ ❯ telnet 45.195.222.93 443
> Connecting To 45.195.222.93...Could not open connection to the host, on port 443: Connect failed
> ```
> 
> From Brazil.

Same. @seydx there is problem from Tuya? Why I got wrong MQTT server from Tuya?

felipecrs commented:
I found an interesting project:

- https://smartathome.co.uk/smartlife/index.html (https://github.com/ndg63276/smartathome)

Could be useful for adding support for Smart Life accounts, maybe.
